posthog/hogql/ast.py

```diff
@@ -16,12 +19,141 @@ class Config:
         method_name = "visit_{}".format(camel_case_name)
-        visit = getattr(visitor, method_name)
-        return visit(self)
+        if hasattr(visitor, method_name):
+            visit = getattr(visitor, method_name)
+            return visit(self)
+        if hasattr(visitor, "visit_unknown"):
+            return visitor.visit_unknown(self)
+        raise ValueError(f"Visitor has no method {method_name}")
+
+
+class Symbol(AST):
+    def get_child(self, name: str) -> "Symbol":
+        raise NotImplementedError("Symbol.get_child not overridden")
+
+    def has_child(self, name: str) -> bool:
+        return self.get_child(name) is not None
+
+
+class FieldAliasSymbol(Symbol):
+    name: str
+    symbol: Symbol
+
+    def get_child(self, name: str) -> Symbol:
+        return self.symbol.get_child(name)
+
+    def has_child(self, name: str) -> bool:
+        return self.symbol.has_child(name)
+
+
+class TableSymbol(Symbol):
+    table: Table
+
+    def has_child(self, name: str) -> bool:
+        return self.table.has_field(name)
+
+    def get_child(self, name: str) -> Symbol:
+        if self.has_child(name):
+            return FieldSymbol(name=name, table=self)
+        raise ValueError(f"Field not found: {name}")
+
+
+class TableAliasSymbol(Symbol):
+    name: str
+    table: TableSymbol
+
+    def has_child(self, name: str) -> bool:
+        return self.table.has_child(name)
+
+    def get_child(self, name: str) -> Symbol:
+        if self.has_child(name):
+            return FieldSymbol(name=name, table=self)
+        return self.table.get_child(name)
+
+
+class SelectQuerySymbol(Symbol):
+    # all aliases a select query has access to in its scope
+    aliases: Dict[str, FieldAliasSymbol] = PydanticField(default_factory=dict)
+    # all symbols a select query exports
+    columns: Dict[str, Symbol] = PydanticField(default_factory=dict)
+    # all from and join, tables and subqueries with aliases
+    tables: Dict[
+        str, Union[TableSymbol, TableAliasSymbol, "SelectQuerySymbol", "SelectQueryAliasSymbol"]
+    ] = PydanticField(default_factory=dict)
+    # all from and join subqueries without aliases
+    anonymous_tables: List["SelectQuerySymbol"] = PydanticField(default_factory=list)
+
+    def get_child(self, name: str) -> Symbol:
+        if name in self.columns:
+            return FieldSymbol(name=name, table=self)
+        raise ValueError(f"Column not found: {name}")
+
+    def has_child(self, name: str) -> bool:
+        return name in self.columns
+
+
+class SelectQueryAliasSymbol(Symbol):
+    name: str
+    symbol: SelectQuerySymbol
+
+    def get_child(self, name: str) -> Symbol:
+        if self.symbol.has_child(name):
+            return FieldSymbol(name=name, table=self)
+        raise ValueError(f"Field not found: {name}")
+
+    def has_child(self, name: str) -> bool:
+        return self.symbol.has_child(name)
+
+
+SelectQuerySymbol.update_forward_refs(SelectQuerySymbol=SelectQuerySymbol)
```

Comment:
As far as I can see, self-references are resolved automatically. https://docs.pydantic.dev/usage/postponed_annotations/#self-referencing-models Is that not true here?

Reply:
Probably true 😅